### ISSUE TITLE
opendistro 1.9, metricsdb files are not cleaned up by default.

### ISSUE THREAD

**Issue by jsirianni** [opendistro-for-elasticsearch/performance-analyzer#155]
Hi, all of my Opendistro systems have very large /tmp directories. It appears that the metricsdb files are not cleaned up. For example, one of my systems has over 60,000 metricsdb files, amounting to around 6GB. 
```
[root@opendistro-master-1 tmp]# ls | grep 'metricsdb_' | wc
  66333   66333 1591992
[root@opendistro-master-1 tmp]# du -sh ./
5.8G	./
[root@opendist
```

It is my understanding that these files should be cleaned up automatically, this is my config (default)
```
[root@opendistro-client-7989bf45ff-mfjd8 /]# cat /usr/share/elasticsearch/performance-analyzer-rca/pa_config/performance-analyzer.properties
# ======================== Elasticsearch performance analyzer plugin config =========================
# NOTE: this is an example for Linux. Please modify the config accordingly if you are using it under other OS.
# Metrics data location
metrics-location = /dev/shm/performanceanalyzer/
# Metrics deletion interval (minutes) for metrics data.
# Interval should be between 1 to 60.
metrics-deletion-interval = 1
# If set to true, the system cleans up the files behind it. So at any point, we should expect only 2
# metrics-db-file-prefix-path files. If set to false, no files are cleaned up. This can be useful, if you are archiving
# the files and wouldn't like for them to be cleaned up.
cleanup-metrics-db-files = true
# WebService exposed by App's port
webservice-listener-port = 9600
# Metric DB File Prefix Path location
metrics-db-file-prefix-path = /tmp/metricsdb_
https-enabled = false
#Setup the correct path for certificates
certificate-file-path = specify_path
private-key-file-path = specify_path
# WebService bind host; default only to local interface
#webservice-bind-host =
# Plugin Stats Metadata file name, expected to be in the same location
plugin-stats-metadata = plugin-stats-metadata
# Agent Stats Metadata file name, expected to be in the same location
agent-stats-metadata = agent-stats-metadata
```

I believe this issue surfaced after upgrading to 1.9. I had 1.8 running stable for weeks.

**Comment by agturley** [comment]
I'm experiencing this same issue with multiple clusters.

**Comment by sendkb** [comment]
Thanks for reporting the issue jsirianni@ and agturley@. We will look into this issue and get back to you.

**Comment by sidheart** [comment]
Hi @jsirianni ,

I was unable to reproduce the issue on a clean installation of opendistro 1.9 on Ubuntu 20.04. It's possible that the new reader process has failed to start on your machine.

One easy way to verify this is to **ssh into the offending node** and issue the following command:

`curl -XGET "localhost:9600/_opendistro/_performanceanalyzer/metrics?metrics=Latency,CPU_Utilization&agg=avg,max&dim=ShardID&nodes=all"`

Do you see a valid response? If not, try searching for the reader process using ps:

`sudo ps aux | grep com.amazon.opendistro.elasticsearch.performanceanalyzer.PerformanceAnalyzerApp`

If nothing is coming back from these commands, you'll have to get the reader process started. Steps 1-3 of [this link](https://opendistro.github.io/for-elasticsearch-docs/docs/install/plugins/#performance-analyzer) should guide you through the process. There are a few differences between how this is done on 1.8 vs 1.9.

I hope this helps you out. If you're still having issues, please share the following details and I'll do my best to reproduce and debug the behavior on my end:

1. Your exact method of installation/upgrade e.g. tarball, rpm distribution, standalone, etc.
1. The output of `ps aux | grep java` on the node
1. The contents of `/tmp/PerformanceAnalyzer.log` on the node
1. The contents of `/var/log/elasticsearch/elasticsearch.log` on the node

**Comment by jsirianni** [comment]
@sidheart 

I have removed the plugin for now, but before removing I was able to use it just fine. I used the example API calls from the README.

1. Install method: Docker image `amazon/opendistro-for-elasticsearch:1.9.1`
2. Not sure that this matters, because I have *uninstalled the plugin*
```
elastic+      69  1.8  7.7 10244052 3583008 ?    Sl   Jul28 211:16 /opt/jdk/bin/java -Xshare:auto -Des.networkaddress.cache.ttl=60 -Des.networkaddress.cache.negative.ttl=10 -XX:+AlwaysPreTouch -Xss1m -Djava.awt.headless=true -Dfile.encoding=UTF-8 -Djna.nosys=true -XX:-OmitStackTraceInFastThrow -XX:+ShowCodeDetailsInExceptionMessages -Dio.netty.noUnsafe=true -Dio.netty.noKeySetOptimization=true -Dio.netty.recycler.maxCapacityPerThread=0 -Dio.netty.allocator.numDirectArenas=0 -Dlog4j.shutdownHookEnabled=false -Dlog4j2.disable.jmx=true -Djava.locale.providers=SPI,COMPAT -Xms1g -Xmx1g -XX:+UseG1GC -XX:G1ReservePercent=25 -XX:InitiatingHeapOccupancyPercent=30 -Djava.io.tmpdir=/tmp/elasticsearch-10566998645916485961 -XX:+HeapDumpOnOutOfMemoryError -XX:HeapDumpPath=data -XX:ErrorFile=logs/hs_err_pid%p.log -Xlog:gc*,gc+age=trace,safepoint:file=logs/gc.log:utctime,pid,tags:filecount=32,filesize=64m -Des.cgroups.hierarchy.override=/ -Xms3072m -Xmx3072m -XX:MaxDirectMemorySize=1610612736 -Des.path.home=/usr/share/elasticsearch -Des.path.conf=/usr/share/elasticsearch/config -Des.distribution.flavor=oss -Des.distribution.type=tar -Des.bundled_jdk=true -cp /usr/share/elasticsearch/lib/* org.elasticsearch.bootstrap.Elasticsearch -Ediscovery.seed_hosts=opendistro-discovery -Enode.data=false -Enode.master=true -Enode.name=opendistro-master-1 -Enetwork.host=0.0.0.0 -Enode.attr.datacenter=us-east1 -Enode.ingest=false
```
3. N/A because I have uninstalled the plugin
4. N/A in a container environment it seems.

I am confident this is an issue with 1.9, because I am running a vanilla image. I did not see this issue on 1.8.

**Comment by sidheart** [comment]
Thanks for the info, I'll work on verifying and diagnosing the behavior.

**Comment by sidheart** [comment]
@jsirianni I've verified the issue and submitted a patch into our master branch. However, the fix won't be present by default until our 1.10 release. For any customer wanting to use Performance Analyzer with the 1.9 Docker image you'll have to use this temporary workaround:

1. Execute a shell session into the container `sudo docker exec -it <container id> /bin/bash`
2. Modify the contents of `performance-analyzer-rca/pa_bin/performance-analyzer-agent`

```
#!/bin/bash

if [ "$START_PERF_ANALYZER" = "false" ]; then
    echo "Not starting performance-analyzer-agent"
    exit 0
fi

if [ -z "$1" ]; then
    if [ -z "$ES_HOME" ]; then
        echo "ES_HOME needs to be set or passed in as the first parameter."
        exit 1
    fi
else
    ES_HOME=$1
fi

if [ -z "$2" ]; then
    if [ -z "$JAVA_HOME" ]; then
        echo "JAVA_HOME needs to be set or passed in as the second parameter."
        exit 1
    fi
else
    JAVA_HOME=$2
fi

if ! echo $* | grep -E '(^-d |-d$| -d |--daemonize$|--daemonize )' > /dev/null; then
    export JAVA_OPTS=-Des.path.home=$ES_HOME\ -Dlog4j.configurationFile=$ES_HOME/performance-analyzer-rca/pa_config/log4j2.xml\ -DconfigFilePath=pa_config/performance-analyzer.properties
    exec $ES_HOME/performance-analyzer-rca/bin/performance-analyzer-rca
else
    echo 'Starting deamon'
    export JAVA_OPTS=-Des.path.home=$ES_HOME\ -Dlog4j.configurationFile=$ES_HOME/performance-analyzer-rca/pa_config/log4j2.xml\ -DconfigFilePath=pa_config/performance-analyzer.properties
    exec $ES_HOME/performance-analyzer-rca/bin/performance-analyzer-rca &

    pid=$!
    PID_LOC=/tmp/performance-analyzer-agent
    if [ -n "$4" ]; then
        PID_LOC=$4
    fi

    if ! ps -p $pid > $PID_LOC ; then
        exit 1
    fi
fi
```

3. Restart the reader process (run `ps aux | grep com.amazon.opendistro.elasticsearch.performanceanalyzer.PerformanceAnalyzerApp`) to get the PID then run `kill -9 <PID>`

This will restore proper cleanup behavior. You'll still need to manually delete the extra metricsdb files that were generated into /tmp.

Thanks for bringing this to our attention.
